cernopendata_client/downloader.py

```diff
@@ -34,15 +40,176 @@
+        total_size = int(response.headers.get("content-length", 0))
+        with open(self.file_dest, self.mode) as f:
+            display_message(
+                msg_type="note",
+                msg="File: ./{}/{}".format(
+                    self.path,
+                    self.file_name,
+                ),
+            )
+            downloaded = self.file_size_offline
+            total_size = total_size + self.file_size_offline
+            for data in response.iter_content(chunk_size=1024):
+                downloaded += len(data)
+                try:
+                    f.write(data)
+                except Exception:
+                    display_message(
+                        msg_type="error",
+                        msg="Download error occured. Please try again.",
+                    )
+                    sys.exit(1)
+                self.show_download_progress(
+                    download_t=total_size, download_d=downloaded
+                )
+                headers = {}
+
+
+class DownloaderHttpPycurl:
+    """Downloader class for managing download related utilities with pycurl downloader engine."""
+
+    def __init__(self, path, file_location, mode, file_size_offline):
+        """Initiate class instance."""
+        self.kb = 1024
+        self.path = path
+        self.mode = mode
+        self.file_location = file_location
+        self.file_name = self.file_location.split("/")[-1]
+        self.file_dest = self.path + "/" + self.file_name
+        self.file_size_offline = file_size_offline if file_size_offline else 0
+
+    def show_download_progress(
+        self, download_t=None, download_d=None, upload_t=None, upload_d=None
+    ):
+        """Show download progress of a file."""
+        download_t = download_t + self.file_size_offline
+        download_d = download_d + self.file_size_offline
+        display_message(
+            msg_type="progress",
+            msg="Progress: {}/{} KiB ({}%)\r".format(
+                str(int(download_d / self.kb)),
+                str(int(download_t / self.kb)),
+                str(int(download_d / download_t * 100) if download_t > 0 else 0),
+            ),
+        )
+        sys.stdout.flush()
+
+    def file_downloader(self):
+        """Download single file with pycurl."""
+        c = pycurl.Curl()
+        c.setopt(c.URL, self.file_location)
+        if self.mode == "ab":
+            c.setopt(c.RESUME_FROM, self.file_size_offline)
+        with open(self.file_dest, self.mode) as f:
+            c.setopt(c.WRITEDATA, f)
+            c.setopt(c.NOPROGRESS, False)
+            c.setopt(c.XFERINFOFUNCTION, self.show_download_progress)
+            try:
+                c.perform()
+            except Exception:
+                display_message(
+                    msg_type="error",
+                    msg="Download error occured. Please try again.",
+                )
+                sys.exit(1)
+            c.close()
+
+
+class DownloaderXrootd:
```

Comment:
The new XRootD downloader does not work locally for me, the troubles are already with checking local file existence before attempting the first download... Here's a simple recipe:

```
$ git clean -d -ff -x
$ $ cernopendata-client download-files --recid 5500  --protocol xrootd
==> Downloading file 1 of 11
Traceback (most recent call last):
  File ".../cernopendata-client/bin/cernopendata-client", line 8, in <module>
    sys.exit(cernopendata_client())
  File ".../cernopendata-client/lib/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File ".../cernopendata-client/lib/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File ".../cernopendata-client/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File ".../cernopendata-client/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File ".../cernopendata-client/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File ".../cernopendata-client/lib/python3.9/site-packages/cernopendata_client/cli.py", line 354, in download_files
    check_error(
  File ".../cernopendata-client/lib/python3.9/site-packages/cernopendata_client/downloader.py", line 211, in check_error
    "size": os.path.getsize(file_dest),
  File "/usr/lib64/python3.9/genericpath.py", line 50, in getsize
    return os.stat(filename).st_size
FileNotFoundError: [Errno 2] No such file or directory: '5500/BuildFile.xml'

```